sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -149,31 +158,141 @@ private[joins] object HashedRelation {
   }
 }
 
+/**
+ * An extended CompactBuffer that could grow and update.
+ */
+private[joins] class MutableCompactBuffer[T: ClassTag] extends CompactBuffer[T] {
+  override def growToSize(newSize: Int): Unit = super.growToSize(newSize)
+  override def update(i: Int, v: T): Unit = super.update(i, v)
+}
 
 /**
  * A HashedRelation for UnsafeRow, which is backed by BytesToBytesMap that maps the key into a
  * sequence of values.
- *
- * TODO(davies): use BytesToBytesMap
  */
 private[joins] final class UnsafeHashedRelation(
     private var hashTable: JavaHashMap[UnsafeRow, CompactBuffer[UnsafeRow]])
   extends HashedRelation with Externalizable {
 
-  def this() = this(null)  // Needed for serialization
+  private[joins] def this() = this(null)  // Needed for serialization
+
+  // Use BytesToBytesMap in executor for better performance (it's created when deserialization)
+  @transient private[this] var binaryMap: BytesToBytesMap = _
+
+  // A pool of compact buffers to reduce memory garbage
+  @transient private[this] val bufferPool = new ThreadLocal[MutableCompactBuffer[UnsafeRow]]
```

Comment:
buffer pool has a special meaning in databases. we should pick a different name.